File: agbac-min/agents/typescript/sender/hybridSender.ts
/**
 * hybridSender.ts
 * Hybrid sender for both in-session and out-of-session agents.
 * 
 * Responsibilities:
 * 1. Extract human `act` data from in-session user session.
 * 2. Provide `act` to in-session agents directly.
 * 3. Securely send `act` to out-of-session agents using TLS + JWT.
 * 4. Include timing and directionality logic for out-of-session transmission.
 */

import fs from 'fs';
import jwt from 'jsonwebtoken';
import https from 'https';
import winston from 'winston';

// Configure logging (match Python style)
const logger = winston.createLogger({
    level: 'info',
    format: winston.format.combine(
        winston.format.timestamp(),
        winston.format.printf(({ timestamp, level, message }) => `${timestamp} [${level}]: ${message}`)
    ),
    transports: [new winston.transports.Console()]
});

// Type for human actor
export interface HumanActor {
    sub: string; // e.g., 'user:alice@example.com'
}

// Type for out-of-session transmission
export interface OutOfSessionTarget {
    url: string;           // Agent endpoint
    tlsCertPath: string;   // Path to TLS certificate for secure communication
    privateKeyPath: string; // Private key to sign JWT
    tokenExpirySec?: number; // Optional JWT expiration
}

export class HybridSender {
    private humanActor: HumanActor;

    constructor(humanActor: HumanActor) {
        this.humanActor = humanActor;
        logger.info(`HybridSender initialized for human actor: ${humanActor.sub}`);
    }

    /**
     * Get human actor data for in-session agents.
     * The in-session agent can directly include this `act` in the token request.
     */
    getInSessionAct(): HumanActor {
        logger.info(`Providing in-session ACT data for: ${this.humanActor.sub}`);
        return this.humanActor;
    }

    /**
     * Send human actor `act` data securely to an out-of-session agent.
     * Uses TLS + signed JWT for authenticity.
     * 
     * @param target - target agent endpoint and TLS/JWT info
     */
    async sendOutOfSessionAct(target: OutOfSessionTarget): Promise<void> {
        logger.info(`Preparing to send out-of-session ACT data for: ${this.humanActor.sub} to ${target.url}`);

        try {
            // Load private key for signing JWT
            const privateKey = fs.readFileSync(target.privateKeyPath, 'utf-8');

            const payload = {
                act: this.humanActor,
                iat: Math.floor(Date.now() / 1000),
                exp: Math.floor(Date.now() / 1000) + (target.tokenExpirySec || 60)
            };

            // Sign JWT
            const token = jwt.sign(payload, privateKey, { algorithm: 'RS256' });

            logger.info(`JWT created for out-of-session transmission`);

            // Load TLS cert for HTTPS request
            const tlsCert = fs.readFileSync(target.tlsCertPath);

            // Send POST request to out-of-session agent securely
            await new Promise<void>((resolve, reject) => {
                const req = https.request(target.url, {
                    method: 'POST',
                    cert: tlsCert,
                    headers: {
                        'Content-Type': 'application/jwt',
                        'Authorization': `Bearer ${token}`
                    }
                }, res => {
                    if (res.statusCode && res.statusCode >= 200 && res.statusCode < 300) {
                        logger.info(`Out-of-session ACT delivered successfully`);
                        resolve();
                    } else {
                        reject(new Error(`Failed to deliver out-of-session ACT: ${res.statusCode}`));
                    }
                });

                req.on('error', (err) => reject(err));
                req.end();
            });
        } catch (error) {
            logger.error(`Error sending out-of-session ACT: ${error.message || error}`);
            throw error;
        }
    }
}

File: agbac-min/agents/python/sender/hybrid_sender.py
import logging
import jwt  # PyJWT library
import time
import requests
from typing import Dict, Optional

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)

class HybridSender:
    """
    Hybrid sender for dual-subject 'act' propagation.
    Supports:
      - In-session agents: pass 'act' data directly.
      - Out-of-session agents: send 'act' data securely using TLS + signed JWT.

    The sender ensures the agent token request will include both:
      - sub (agent identity)
      - act (human identity)
    """

    def __init__(self, private_key_pem: Optional[str] = None, tls_verify: bool = True):
        """
        :param private_key_pem: PEM-formatted private key for signing JWTs (out-of-session)
        :param tls_verify: whether to verify TLS certificates for out-of-session requests
        """
        self.private_key_pem = private_key_pem
        self.tls_verify = tls_verify

    def prepare_in_session_act(self, human_identity: Dict) -> Dict:
        """
        Prepare 'act' data for an in-session agent.
        The human identity is directly provided to the agent code.
        """
        logger.info(f"Preparing in-session 'act' data: {human_identity}")
        return human_identity

    def prepare_out_of_session_act(self, human_identity: Dict, agent_endpoint: str, ttl_seconds: int = 60) -> str:
        """
        Prepare 'act' data for an out-of-session agent.
        Uses a signed JWT to securely convey the human identity.
        
        :param human_identity: The 'act' data of the human user
        :param agent_endpoint: URL of the out-of-session agent to receive the JWT
        :param ttl_seconds: Time-to-live for the JWT
        :return: JWT string
        """
        if not self.private_key_pem:
            raise ValueError("Private key required for out-of-session JWT signing.")

        now = int(time.time())
        payload = {
            "act": human_identity,
            "iat": now,
            "exp": now + ttl_seconds,
            "aud": agent_endpoint
        }

        try:
            signed_jwt = jwt.encode(payload, self.private_key_pem, algorithm="RS256")
            logger.info(f"Signed JWT prepared for out-of-session agent at {agent_endpoint}")
            return signed_jwt
        except Exception as e:
            logger.error(f"Failed to generate JWT for out-of-session agent: {e}")
            raise

    def send_out_of_session_act(self, signed_jwt: str, agent_endpoint: str) -> Dict:
        """
        Send the signed JWT to the out-of-session agent over HTTPS.
        :param signed_jwt: The signed JWT containing 'act' data
        :param agent_endpoint: URL of the out-of-session agent
        :return: HTTP response JSON
        """
        headers = {"Authorization": f"Bearer {signed_jwt}"}
        try:
            response = requests.post(agent_endpoint, headers=headers, verify=self.tls_verify)
            response.raise_for_status()
            logger.info(f"'act' data successfully sent to out-of-session agent at {agent_endpoint}")
            return response.json()
        except requests.HTTPError as e:
            logger.error(f"HTTP error sending 'act' to out-of-session agent: {e}")
            raise
        except Exception as e:
            logger.error(f"Unexpected error sending 'act' to out-of-session agent: {e}")
            raise
